### PR TITLE
CORTX-29024: Disable left shift of K:V pairs when lower indexes are deleted in a l…

### PR DESCRIPTION
**Pull request by suryakumar1024**
…ist/array

# Problem Statement
- Existing conf store delete will do left shift k,v if any lower indexes are deleted. which makes the indexes to be a random and less confident for the external component to fetch data based on the index.

# Design
-  This patch will remove the left shift approach by placing none as value if any lower indexes got deleted.
- The last index of any list will be deleted (We don't add any none here ) if user calls delete explicitly 

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: 
-  Confirm All CODACY errors are resolved [Y/N]: 

# Testing 
- [x] Confirm that Test Cases are added (for both the cases, fix and feature) [Y/N]: N
- [x] Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: NA
- [x] Confirm Testing was performed with installed RPM [Y/N]:  Y

# Review Checklist 
  Before posting the PR please ensure
- [x] PR is self reviewed
- [x] Is there a change in filename/package/module or signature [Y/N]: Y
- [x] If yes for above point, Is a notification sent to all other cortx components [Y/N]: Y
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [x] All the tests performed should be mentioned before Resolving a JIRA. 
- [ ] Verification needs to be done before marked as Closed/Verified 

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

**Review by selvakumaar5496 — APPROVED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 1]
```diff
@@ -281,10 +281,7 @@ def get(index: str, key: str, default_val: str = None, **filters):
     @staticmethod
     def delete(index: str, key: str):
         """ Deletes a given key from the config """
-        is_deleted = Conf._conf.delete(index, key)
-        if is_deleted:
-            Conf.save(index)
-        return is_deleted
+        return Conf._conf.delete(index, key)
```
Comment (rahul-tripathi-git) [inline py-utils/src/utils/conf_store/conf_store.py:284]: Why is saving back to source file disabled?

Or can we add a flag for this, to let the user select if he wish to make the delete permanent?
CC: @ujjwalpl

File: py-utils/src/utils/conf_store/conf_store.py [hunk 2]
```diff
@@ -342,7 +339,6 @@ def search(index: str, parent_key: str, search_key: str,
     def add_num_keys(index):
         """Add "num_xxx" keys for all the list items in ine KV Store."""
         Conf._conf.add_num_keys(index)
-        Conf.save(index)
```
Comment (rahul-tripathi-git) [inline py-utils/src/utils/conf_store/conf_store.py:345]: same goes for here.

**Comment by rahul-tripathi-git** [comment]
Hi @suryakumar1024 
Patch looks good. Please ass a few test cases including the corner cases.

File: py-utils/src/utils/kv_store/kv_payload.py [hunk 3]
```diff
@@ -344,7 +344,10 @@ def _delete(self, key: str, data: dict):
             if index >= len(data[k[0]]):
                 return False
             if len(k) == 1:
-                del data[k[0]][index]
+                if index == len(data[k[0]]) - 1:
```
Comment (rohit-k-dwivedi) [inline py-utils/src/utils/kv_store/kv_payload.py:347]: any test addition or previous change required with these change?

**Review by ujjwalpl — APPROVED** [review]

File: py-utils/src/utils/kv_store/kv_payload.py (hunk 3 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kv_payload.py:347]: Added new test cases.

Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kv_payload.py:347]: No changes required for previous test cases

**Review by rohit-k-dwivedi — APPROVED** [review]

File: py-utils/test/kv_store/test_kv_store.py [hunk 4]
```diff
@@ -434,6 +434,28 @@ def test_010_dict_store_set_empty_value_to_key(self):
...
+        result_list_elements = result_list[0]
+        self.assertEqual(len(initial_list_elements),len(result_list_elements))
+
+    def test_no_left_shift_end_list_kvstore(self):
+        """
+        Test and ensure left shift not happening
+        while deleting last element of a list
+        """
+        TestStore.loaded_json[0].set(['bridge>end_name_list'],\
+            [['1', '2', '3', '4']])
```
Comment (rahul-tripathi-git) [inline py-utils/test/kv_store/test_kv_store.py:454]: Please set it one by one .
Conf.set(str, str)

**Review by rahul-tripathi-git — CHANGES_REQUESTED** [review]

File: py-utils/test/kv_store/test_kv_store.py (hunk 4 above)
Comment (suryakumar1024) [reply py-utils/test/kv_store/test_kv_store.py]: Updated test cases.

File: py-utils/test/kv_store/test_kv_store.py [hunk 5]
```diff
@@ -434,6 +434,27 @@ def test_010_dict_store_set_empty_value_to_key(self):
...
+    def test_no_left_shift_end_list_kvstore(self):
+        """
+        Test and ensure left shift not happening
+        while deleting last element of a list
+        """
+        TestStore.loaded_json[0].set(['bridge>end_name_list[0]',\
+            'bridge>end_name_list[1]', 'bridge>end_name_list[2]',\
+            'bridge>end_name_list[3]'], ['1', '2', '3', '4'])
+        TestStore.loaded_json[0].delete(['bridge>end_name_list[3]'])
+        result_list = TestStore.loaded_json[0].get(['bridge>end_name_list'])
```
Comment (rahul-tripathi-git) [inline py-utils/test/kv_store/test_kv_store.py:455]: Do not get a list .This functionality might get removed in near future.

Comment (suryakumar1024) [reply py-utils/test/kv_store/test_kv_store.py]: Updated test cases

**Review by rahul-tripathi-git — APPROVED** [review]
+1
